---
title: "Gene-expression analysis with RNA sequence data using R"
date: "4/12/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE)
```

## Presenters
Huiwen Zheng (AIBN, UQ)

Atefeh Taherian Fard (AIBN, UQ)

## Resources and data files 
You can access and download the dataset used in the workshop through GEO website with accession number: [GSE65267](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65267) 

**Note**: Please download the following files and place them in your`./data` directory.

Data files:

* Metadata.csv
* Data.csv

## Overview

* Loading gene expression data into the R environment
* Identifying and filtering lowly expressed genes
* Differential expression analysis
* Normalisation
* Principal Component Analysis (PCA)
* Visualisation
* Pathway over-representation analysis

## Introduction and data import 

The dataset used in this tutorial contains the RNA-sequence data of mouse kidney before and at various time points (1, 2, 3, 7 & 14 days) after intraperitoneal treatment with folic acid to investigate kidney function-loss.

There are many steps involved in analysing RNA-seq experiment which begins with sequencing reads. The reads are aligned to a reference genome; then the number of reads mapped to each gene is counted, resulting in a table of counts, which is what we use to perform statistical analyses on in R today.

*Set up an RStudio project specifying the directory where you have saved the `/data` directory*

First, let's load all the packages we need to anlyse the data.

```{r, message = FALSE}
#install.packages("scales", repos = 'http://cran.us.r-project.org')
#install.packages("pheatmap", repos = 'http://cran.us.r-project.org')
#install.packages("BiocManager") #to install and manage packages in Bioconductor 

#BiocManager::install(c("DESeq2","org.Mm.eg.db", "vsn"))


library(DESeq2)
library(org.Mm.eg.db)
library(RColorBrewer)
library(clusterProfiler)
library(pheatmap)
```


### Dataset-Mouse chronic kidney disease

The data used in this tutorial comes from the following paper [RNA Sequencing Identifies Novel Translational Biomarkers of Kidney Fibrosis](https://www.ncbi.nlm.nih.gov/pubmed/26449608)

### Loading the count data

```{r}
#Load the data into R
kidney.dat <- read.csv("data/Count_matrix.csv", row.names = 1, stringsAsFactors = F)

#Load the sample information (metadata) into R
metadata <- read.csv("data/Metadata.csv", row.names = 1, stringsAsFactors = F)

rownames(metadata) <- colnames(kidney.dat) # Make sure the columns in the dataset are matching to the row names in the metadata file 

# Examine the count matrix
head(kidney.dat)
#str(kidney.dat)

# Examine the metadata file
head(metadata)

boxplot(kidney.dat, main = "Raw Counts", las = 2) # A quick look at the data 
#use par() to specific set/increase the plot's window par(mar = c (10,10,4,2)+0.1)

# Create DESeqDataSet
dds <- DESeqDataSetFromMatrix(countData = kidney.dat,
                              colData = metadata,
                              design = ~ group)
dds

```

### Filtering out low expressed genes

Here we perform a minimal pre-filtering to keep genes that are expressed in atleast 10% in the total number of the samples. Total of 23888 genes remain after this filtering process.
*Note:* It is not necessary to pre-filter low count genes before running DESeq2 function. Reasons to perform this step are 1) to increase the speed of transformation sand testing and 2) to reduce the memory size of the dds.

```{r}
keep <- rowSums(counts(dds)) >= round(0.1*length(kidney.dat))
length(keep) # Numbers of samples that have more than 10 genes expressed
dds <- dds[keep,]
length(dds)
```

### Differential gene expression analysis

The standard differential expression analysis steps are wrapped into a single function. DESeq.Results tables are generated using the function results, which extracts a results table with log2 fold changes, p-values and adjusted p-values.

If we don't specify which group we want to compare, the comparsion will be based on the alphabatic order of the levels AND the foldchange in the result table will be the last level of this variable over the reference level.

*Note:* You can change the reference level using the relevel() function. 


```{r}
# So let's see what are the levels in the group object
factor(metadata$group)
dds <- DESeq(dds)

# Results tables are generated using the function results()
res0 <- results(dds)
res0 # Comparing between Day7 vs Control
```

In the result table, the lfcSE gives the standard error of the log2FoldChange. For the Wald test, stat is the Wald statistic: the log2FoldChange divided by lfcSE, which is compared to a standard Normal distribution to generate a two-tailed pvalue.

If you want to choose a specific pairwise comparison, you can use `name` or `contrast` function to make your own contrast list. 

Details about the comparison are printed to the console, directly above the results table. The text, condition Day 1 vs Day 14, tells you that the estimates are of the logarithmic fold change log2(Day7/Control).

```{r}
#res <- results(dds, name="group_Day1_vs_Control")
res <- results(dds, contrast=c("group","Day1","Control"))
res

```

### P-values and adjusted p-values

Note that the results function automatically performs independent filtering based on the mean of normalised counts for each gene, optimising the number of genes with adjusted p-value below a given FDR cutoff, `alpha` (The default is 0.1).

How many adjusted p-values are smaller than 0.05?

```{r}
res05 <- results(dds, alpha=0.05)

# Counting the number of significantly expreseed genes with alpha at 0.05
sum(res05$padj < 0.05, na.rm=TRUE)

```

### Further exploring DE genes
We make a volcano plot based on the DE analysis results between **Day1** & **Control**, p.adj values and log2FoldChanges. Genes that have p.adj<0.05 are colored in red.

```{r}
# Make a basic volcano plot
with(res, plot(log2FoldChange, -log10(padj), pch=20, main="Volcano plot", xlim=c(-10,10), ylim=c(0,50)))

# Add colored points: red if padj<0.05
with(subset(res, padj<.05 ), points(log2FoldChange, -log10(padj), pch=20, col="red"))

```

### Data transformation for clustering and visualisation purposes 

We perform differential expression analysis on raw counts; this is because data normalisation in the DESeq2 model happens internally, correcting for library size. However for other downstream analyses - e.g. for visualisation or clustering - it is useful to work with the transformed/scaled versions of the count data.

There are two data transformation methods for count data in Deseq2: **variance stabilizing transformations (VST)** and **regularized logarithm (rlog) **. 

Here we use to *VST* for transforming the data.

```{r}
vsd <- vst(dds) 
#rld <- rlog(dds)

head(assay(vsd),5) # Have a look at the normalised gene expression
```

### The effect of normalisation

The figure below plots the standard deviation of the transformed data, across samples, against the mean, using vst.

```{r}
library("vsn")
meanSdPlot(assay(vsd))
#meanSdPlot(assay(rld))
```

The figure below shows the expression distribution of samples before and after normalisation using **VST**. The data is normalised by `Median of ratios method`.

```{r}
library(RColorBrewer)
col_vector<-c('#e6194b', '#3cb44b', '#ffe119', '#4363d8', '#f58231', '#911eb4', '#46f0f0', '#f032e6', '#bcf60c', '#fabebe', '#008080', '#e6beff', '#9a6324', '#fffac8', '#800000', '#aaffc3', '#808000', '#ffd8b1')

# this gives log2(n + 1) transformation only
ntd <- normTransform(dds)
boxplot(assay(ntd), las=2,col=col_vector)
title(main="Example 1: Not transformed",ylab="Log-cpm")

boxplot(assay(vsd), las=2,col=col_vector)
title(main="Example 2: Transformed",ylab="Log-cpm")

```

### Dimensional Reduction visualisation 

**Principal component plot** is used for visualising the overall effect of experimental covariates.

```{r}
plotPCA(vsd, intgroup = "group")
#plotPCA(rld, intgroup = "group")

```

To explore the significantly expressed genes, it is often informative to plot a heatmap. Below we show how to create a heatmap of DE genes in all groups.

```{r}
# First we order the genes based on their p.adj values and identify the top 100 for plotting the heatmap

resOrdered <- res[order(res$padj),]# Order the dataframe based on the adjusted p-values
df <- assay(vsd)[rownames(assay(vsd)) %in% resOrdered@rownames[1:100],] # Select the top 100 genes
df <- scale(df) # Scale the data before plotting   

anno <- as.data.frame(colData(dds)[,"group"])
rownames(anno) <- colnames(dds)
colnames(anno) <- "group"

pheatmap(df, cluster_rows= TRUE, show_rownames=FALSE,
         cluster_cols=FALSE, annotation_col = anno)

```



### Pathway over-representation anaysis

Over-representation methods compare sets of genes annotated to pathways to a list of those genes that are significantly deferentially expressed between two phenotypes. Here we use [Gene Ontology Database](http://geneontology.org/). 

```{r}
# We extract 500 genes for over-representation analysis
Genelist <- resOrdered@rownames[1:500]

# We use BP -- represents Biological Processes for downstream analysis 
orgo <- enrichGO(Genelist, OrgDb = org.Mm.eg.db, ont='BP',pvalueCutoff = 0.05, qvalueCutoff = 0.2,keyType = 'SYMBOL')

dotplot(orgo,showCategory=10) # Dotplot for the top 10 GO terms 
cnetplot(orgo) # Network plot of significant terms
```


Pathway Over-representation analysis using Hallmark gene signatures. Hallmark dataset can be found here: [Hallmark](http://software.broadinstitute.org/gsea/msigdb/index.jsp)


```{r}
#install.packages("msigdbrs")
library(msigdbr)
hallmark <- msigdbr(species = "Mus musculus", category = "H")
poa <- enricher(Genelist, TERM2GENE=hallmark[c("gs_name", "gene_symbol")])

dotplot(poa) # Dotplot for the top 10 enrihced pathways 
cnetplot(poa) # Network plot of significant terms
```

### References and other resources

* [DESeq2](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html)
* [ClusterProfiler](https://yulab-smu.github.io/clusterProfiler-book/)
* [RNAseq123](https://www.bioconductor.org/packages/release/workflows/html/RNAseq123.html)
* [Bioconductor workflows](https://www.bioconductor.org/packages/release/BiocViews.html#___Workflow)
* [Reproducibilty and version control](https://github.com/)













